### DISCUSSION TITLE
Duckdb update is 4x slower than Pandas

### DISCUSSION THREAD

**Discussion by zachary62** [duckdb/duckdb#3492] [Q&A]
In your post https://duckdb.org/2021/05/14/sql-on-pandas.html, DuckDB is a nice replacement to pandas for read-only queries.

I have some data analytics workloads that not only query data, but write large intermediate results. I find that duckdb is 4x slower than pandas. Consider the following task:

```import pandas as pd
import duckdb
R = pd.DataFrame()
R["A"] = range(10000000)
R["B"] = range(10000000)
R.to_csv("example.csv", index=False)
con = duckdb.connect(database=':memory:')
con.execute("CREATE TABLE R AS SELECT * FROM 'example.csv';")
```

DuckDB is 2x faster for aggregation
```
# 7 ms
R["B"].sum()
# 3 ms
print(con.execute("SELECT SUM(B) FROM R;"))
```

However, when I want to write intermediate results, update is 4x slower
```
# 52 ms
R["A"] = R["A"] - R["B"]
# 230 ms
print(con.execute("UPDATE R SET A=A-B;"))
```

This is fine as all OLAP db are slow in updates.

Just wondering where is the overhead from? Is it from mvcc? What if I don't care about acid for some table? Can I turn off mvcc and get fast update performance?

**Comment by zachary62** [comment]
Any update to this question?
Is it from concurrency or compression?

**Reply by Mytherin** [reply]
There are various reasons. In general updates in DuckDB are far more complex than updates in Pandas.

* Pandas performs an update on uncompressed data that is always residing entirely in memory, in a single contiguous array, with no concern for partial updates or data loss.
* DuckDB performs an update on data that may or may not reside in memory, that may or may not be compressed, that is partitioned over multiple blocks in storage, and performs updates with support for ACID semantics using MVCC, while also having to handle the case of multiple concurrent readers and writers accessing the table.

All things considered four times slower does not seem that bad. For reference, performing that same update in SQLite takes >50x longer than Pandas and >10x longer than DuckDB on my machine.

There are ways we could speed up updates further in the future. For example, update queries run single threaded currently. We could provide support for turning off MVCC. But those features are not there (yet), and not a very high priority right now.

**Reply by zachary62** [reply]
Hi @Mytherin, thank you very much for your response!
Just to better break down the factors, I also redo experiment with disk-based duckDB, and update by creating a new table. (results are different from previous ones maybe because system issues)
Here are my updated experiments:

```
import pandas as pd  
import duckdb  
example_pandas = pd.DataFrame()  
example_pandas["A"] = range(10000000)  
example_pandas["B"] = range(10000000)  
example_pandas.to_csv("example.csv", index=False)
con_mem = duckdb.connect(database=':memory:')
con_mem.execute("CREATE OR REPLACE TABLE example AS SELECT * FROM 'example.csv';")
con_disk = duckdb.connect(database='exp.duckdb')
con_disk.execute("CREATE OR REPLACE TABLE example AS SELECT * FROM 'example.csv';")
```

Here are update performance:
```
# 47.4ms
example_pandas["A"] = example_pandas["A"] - example_pandas["B"]

# 317ms
con_mem.execute("UPDATE example SET A=A-B;")

# 817ms
con_disk.execute("UPDATE example SET A=A-B;")

# instead update in-place, create a new table
# 93.8ms
con_mem.execute("CREATE OR REPLACE TABLE example AS SELECT A-B FROM example;")

# 450ms
con_disk.execute("CREATE OR REPLACE TABLE example AS SELECT A-B FROM example;")
```

There are many factors: compression, MVCC concurrency, WAL, data partitioned across memory and disks, and single thread.

Many factors are disk-only: MVCC concurrency, WAL, data partitioned across memory and disks. 
That's the gap between 817ms and 317 ms?

I assume that, for a single process running in-memory duckdb, there should be no concurrency concern?
And all data are only in memory.
The gap between in-memory duckDB (317ms) and Pandas (47.4ms) is mainly because it is writing to the compressed data. And data may be partitioned (even if it's all in memory)?

Pandas is only single-threaded, so the duckdb single-thread update doesn't cause the gap here.
 
Pandas doesn't update in place. It creates a new column.
When DuckDB creates a new table, it is faster. This is because the overhead of writing to compressed data is removed.
There is still a gap 47.4ms vs 93.8ms. This is the overhead of creating the table and doing the compression?

Really appreciate your insights!

**Reply by Alex-Monahan** [reply]
Today, an update is a delete and then an insert I believe. That is why a brand new table is faster I imagine!

**Reply by zachary62** [reply]
@Alex-Monahan For column store, is delete + insert slow? 
Let's say we change a single cell. Do you need to delete the whole row from other columns, then insert a row which also insert to all other columns?
Also, how do you recompress?

**Reply by Mytherin** [reply]
MVCC and concurrency are not disk only. DuckDB supports opening multiple connections to an in-memory database, and MVCC is required even for single-connection workloads in order to support ROLLBACK and all-or-nothing transaction semantics.

**Reply by zachary62** [reply]
I see. Thank you very much!
So to sum up:
- 817ms -> 317ms: WAL and querying data across memory/disks
- 317ms -> 93.8ms: Update is not optimized. It's delete+insert. (update has a larger ACID overhead?)
- 93.8ms -> 47.4ms: Table overhead. Compression. ACID.
